### PR TITLE
Mekhanik evgenii/fix 1995 5

### PR DESCRIPTION
**Pull request by EvgeniiMekhanik**

### CODE REVIEW

**Comment by EvgeniiMekhanik** [comment]
DRAFT until 1995.1-3 will be merged

File: fw/cache.c [hunk 1]
```diff
@@ -1117,10 +1124,11 @@ tfw_cache_entry_key_eq(TDB *db, TfwHttpReq *req, TfwCacheEntry *ce)
 	 * In contrast, a POST request cannot be satisfied by a cached
 	 * POST response because POST is potentially unsafe.
 	 */
-	if (req->method == TFW_HTTP_METH_POST
-	    && ce->method == TFW_HTTP_METH_POST)
+	if ((req->method == TFW_HTTP_METH_POST
```
Comment (const-t) [inline fw/cache.c:1127]: Double brackets is unnecessary here.

File: fw/cache.c [hunk 2]
```diff
@@ -3257,9 +3296,10 @@ tfw_cfgop_cache_methods(TfwCfgSpec *cs, TfwCfgEntry *ce)
 			return -EINVAL;
 		}
 		if (__cache_method_nc_test(method)) {
-			T_WARN_NL("%s: non-cacheable method '%s' is set "
+			T_ERR_NL("%s: non-cacheable method '%s' is set "
 				  "as cacheable\n",
 				  cs->name, val);
+			return -EINVAL;
```
Comment (const-t) [inline fw/cache.c:3302]: It totally breaks our [cache_methods](https://github.com/tempesta-tech/tempesta/wiki/Caching-Responses#cache_methods). We must discuss it.

File: fw/cache.c [hunk 3]
```diff
@@ -312,7 +313,10 @@ __cache_method_test(tfw_http_meth_t method)
 static inline bool
 tfw_cache_msg_cacheable(TfwHttpReq *req)
 {
-	return cache_cfg.cache && __cache_method_test(req->method);
+	/* POST request is not idempotent, but can be cacheble. */
+	return cache_cfg.cache && __cache_method_test(req->method) &&
+		(!tfw_http_req_is_nip(req)
+		 || req->method == TFW_HTTP_METH_POST);
```
[the remark this answers is not in the document]
Comment (EvgeniiMekhanik) [reply fw/cache.c:319]: See task https://github.com/tempesta-tech/tempesta/issues/1663 and comment to it

**Comment by EvgeniiMekhanik** [comment]
I switch this PR to draft until https://github.com/tempesta-tech/tempesta/pull/2074/files#diff-6e5c29ba3006ec5b4a6736951be30dd7826be9f261e66a33119efb5bab43dfd7 will be merged. After discussion we decide that we should not invalidate all cache records for successful GET and HEAD responses. We should save new records in cache and take first appropriate record (which should be the youngest one). Currently (on master we get first the oldest one!). It will be implemented in 2074.

**Comment by EvgeniiMekhanik** [comment]
According our investigations nginx doesn't cache any other methods beside GET/HEAD/POST, so we can disable other methods from cache

File: fw/cache.c (hunk 2 above)
Comment (EvgeniiMekhanik) [reply fw/cache.c:3289]: After our discussion we decide to do it in this way

**Review by krizhanovsky — APPROVED** [review]
I don't anything to add to @const-t review, so LGTM for me

**Review by const-t — APPROVED** [review]

**Pull request merged**